enterprise/server/invocation_stat_service/invocation_stat_service.go

```diff
@@ -374,6 +400,264 @@ func (i *InvocationStatService) GetInvocationStatBaseQuery(aggColumn string) str
+	width := valueRange.High + 1 - valueRange.Low
+	var step int64
+
+	bucketCount := int64(20)
+	if width <= bucketCount {
+		step = 1
+	} else {
+		step = width / bucketCount
+		if width%bucketCount != 0 {
+			step = step + 1
+		}
+	}
+
+	return valueRange.Low, step, bucketCount, nil
+}
+
+func (i *InvocationStatService) getMetricBuckets(ctx context.Context, table string, metric string, whereClauseStr string, whereClauseArgs []interface{}) ([]int64, string, error) {
+	metricLow, metricBucketSize, metricBucketCount, err := i.getMetricRange(ctx, table, metric, whereClauseStr, whereClauseArgs)
+	if err != nil {
+		return nil, "", err
+	}
+	if metricBucketCount == 0 {
+		return nil, "", nil
+	}
+
+	metricBuckets := make([]int64, metricBucketCount+1)
+	for i := range metricBuckets[:] {
+		metricBuckets[i] = metricLow + int64(i)*metricBucketSize
+	}
+	mStr := make([]string, metricBucketCount)
+	for i := range mStr {
+		mStr[i] = fmt.Sprint(metricBuckets[i])
+	}
+	metricArrayStr := "array(" + strings.Join(mStr, ",") + ")"
+	return metricBuckets, metricArrayStr, nil
+}
+
+func getTimestampBuckets(q *stpb.TrendQuery, timezoneOffsetMinutes int32) ([]int64, string, error) {
+	lowTime := q.GetUpdatedAfter()
+	highTime := q.GetUpdatedBefore()
+	if lowTime != nil && !lowTime.IsValid() {
+		return nil, "", status.InvalidArgumentError(fmt.Sprintf("Invalid low timestamp: %v", lowTime))
+	}
+	if highTime != nil && (!highTime.IsValid() || highTime.AsTime().Unix() < lowTime.AsTime().Unix()) {
+		return nil, "", status.InvalidArgumentError(fmt.Sprintf("Invalid high timestamp: %v", highTime))
+	}
+
+	offsetSec := int64(timezoneOffsetMinutes) * 60
+
+	startSec := time.Now().Unix() - ((7 * 24 * time.Hour).Microseconds() / 1e6)
+	if lowTime != nil {
+		startSec = lowTime.GetSeconds()
+	}
+
+	endSec := time.Now().Unix()
+	if highTime != nil {
+		endSec = highTime.GetSeconds()
+	}
+
+	// TODO(jdhollen): Bucket at true midnight in user timezone instead of fudging
+	// it, this gets hosed around daylight savings transitions.  This matters way
+	// less for hourly buckets (where we can easily let the client format).
+	s := time.Unix(startSec, 0).UTC().Add(-time.Minute * time.Duration(timezoneOffsetMinutes))
+	localStart := time.Date(s.Year(), s.Month(), s.Day(), 0, 0, 0, 0, s.Location())
+	currentDateBracket := localStart.Add(time.Minute * time.Duration(timezoneOffsetMinutes))
+
+	currentString := currentDateBracket.Format("2006-01-02")
+	endString := time.Unix(endSec-offsetSec, 0).UTC().AddDate(0, 0, 2).Format("2006-01-02")
+
+	var timestampBuckets []int64
+	for currentString != endString {
+		timestampBuckets = append(timestampBuckets, currentDateBracket.UnixMicro())
+		currentDateBracket = currentDateBracket.AddDate(0, 0, 1)
+		currentString = currentDateBracket.Format("2006-01-02")
+	}
+
+	dStr := make([]string, len(timestampBuckets)-1)
+	for i, d := range timestampBuckets[:len(timestampBuckets)-1] {
+		dStr[i] = fmt.Sprintf("%d", d)
+	}
```

Comment:
big nit but you could just define a var for `numDateBuckets = len(timestampBuckets)-1` then use that instead of len(timestampBuckets)-1 in multiple places so folks don't have to squint at the range indexing as hard

Reply:
Sure, why not.